[BE] 배포 자동화 파이프라인 리팩터링

YehyeokBang opened the pull request:
<!-- PR 제목 예시: [AN] 로그인 화면 UI 구현 / [BE] 회원가입 API 추가 / [ALL] 공통 유틸 정리 -->

## ✅ 작업 개요
<!-- 어떤 기능/수정/리팩토링인지 한 줄로 설명 (예: 로그인 화면 UI 구현) -->

- 추후 적용할 무중단 배포 파이프라인을 위한 배포 환경 개선
- 무중단 배포 적용이 아닙니다. 기존 돌아가기만 하던 개발 서버 배포 과정을 개선한 내용입니다.

<br>

## 🎯 관련 이슈
<!-- 관련된 이슈 번호를 태깅해주세요. -->
- Closed #586 

<br>

## 🔍 상세 내용
<!-- 
작업한 내용을 구체적으로 작성해주세요.
- 로그인 화면 UI 구성 (이메일, 비밀번호 입력란, 로그인 버튼)
- 입력값 유효성 검사 추가
- ViewModel 및 상태 관리 로직 연동
-->

<img width="1082" height="427" alt="image" src="https://github.com/user-attachments/assets/4d241100-d006-461f-b08e-4aa5de216cd5" />

[Wiki 최신화](https://github.com/woowacourse-teams/2025-bottari/wiki/%EA%B0%9C%EB%B0%9C-%EC%84%9C%EB%B2%84-%EB%B0%B0%ED%8F%AC-%EC%9E%90%EB%8F%99%ED%99%94-%ED%8C%8C%EC%9D%B4%ED%94%84%EB%9D%BC%EC%9D%B8)도 미리 진행했어요!

자세한 내용은 리뷰로 남겨놓을게요.

- 개발 서버 내에 `docker-compose-dev.yml` 파일 추가
- `deploy-dev-server.yml` 파이프라인 수정
- 프로젝트 루트에 `Dockerfile` 추가

<br>

## 💬 기타 참고사항
<!-- 
코드 리뷰어가 참고해야 할 사항이 있다면 적어주세요.
- 추후 `회원가입` 화면에서도 재사용 가능한 UI 컴포넌트로 분리할 예정입니다.
- `AuthViewModel` 내 중복 로직은 이후 리팩토링에서 정리 예정입니다.
-->

<br>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * 백엔드에 멀티스테이지 컨테이너 빌드를 도입해 빌드와 실행 이미지를 분리했습니다.
  * 실행 이미지에 빌드 산출물을 포함하고 8080 포트를 노출하며 기본 표준시간대 설정을 추가했습니다.
  * 패키징 방식이 변경되어 실행 가능한 애플리케이션 단일 배포 파일로 빌드되도록 조정했습니다.
  * 이미지 크기 감소와 함께 배포 효율·시작 시간 및 운영 안정성이 개선되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

YehyeokBang committed:
refactor: 배포 자동화 파이프라인 수정

- 도커 기반 환경으로 변경

YehyeokBang commented on backend/bottari/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+# ===== Build Stage =====
+FROM eclipse-temurin:21-jdk-alpine AS build
+WORKDIR /app
+COPY build.gradle settings.gradle gradlew /app/
+COPY gradle /app/gradle
+COPY src /app/src
+RUN chmod +x gradlew && ./gradlew clean build -x test
+
+# ===== Runtime Stage =====
+FROM eclipse-temurin:21-jre-alpine
+WORKDIR /app
+COPY --from=build /app/build/libs/*.jar /app/app.jar
+EXPOSE 8080
+ENTRYPOINT ["java", "-Duser.timezone=Asia/Seoul", "-jar", "/app/app.jar"]
```
> 요약: Dockerfile로 이미지를 만들 때, `빌드할 때 필요한 도구`는 결과물에 포함시키지 않도록 하여 `용량을 최소화 함`

크게 2단계로 볼 수 있는 멀티 스테이지 빌드(Multi-stage Build) 구조 채택

1. Build Stage
    - Gradle로 JAR 파일을 빌드하는 단계
    - JDK(개발용) 이미지 사용
2.	Runtime Stage
	- 완성된 JAR만 복사해서 실행하는 단계
	- JRE(런타임 전용) 이미지 사용

최종 실행 이미지는 오직 `JRE + app.jar`만 포함시킴
→ 빌드 도구, Gradle 캐시 등은 포함되지 않아 `용량 최소화`

YehyeokBang commented on .github/workflows/deploy-dev-server.yml:
```diff
@@ -2,14 +2,23 @@ name: deploy-dev-server
 
 jobs:
   build:
+    name: Build & Push to GHCR
     runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write  # GHCR push를 위해 필요
```
## GHCR

이미지를 저장할 수 있는 저장소

## Docker Hub, ECR 이런건???

- 현재 인프라가 GitHub Actions를 중심으로 구성되어 있고, GHCR은 같은 GitHub 생태계 안에서 `무료`임
- AWS ECR은 추후 확장성 면에서 유리하지만, 서비스 규모가 크지 않고, `우테코 캠퍼스 지원이 종료될 경우 AWS 자체를 유지하지 못할 가능성`도 있음
- docker-hub는 예전엔 편하게 무료였지만, 점차 유료로 변경되가며, pull 횟수 제한도 있다고 들었는데 흠..

간편하고 무료이며, `특정 클라우드에 대한 의존성이 적은 GHCR을 사용`하는 것이 유연한 선택이라고 판단함

jaehyeon2650 replied:
바로 물어보려고 했는데 답변 예샷

YehyeokBang commented on .github/workflows/deploy-dev-server.yml:
```diff
@@ -2,14 +2,23 @@ name: deploy-dev-server
 on:
   push:
-    branches: be/develop
+    branches:
+      - be/develop
```
## 흐름

1. `be/develop`에 변경 사항 push
2. 우분투 환경에서 도커 이미지를 빌드하고 `GHCR(깃허브 이미지 저장소)`에 push
3. `self-hosted`로 이어받아서 `GHCR`에서 이미지 pull
4. EC2에 인스턴스에 있는 `docker-compose-dev.yml` 파일 실행 (+환경변수 주입)

## 고민 지점

> 일단은 편의성을 위해 개발 서버에서는 인스턴스 내부에서 docker-compose 파일을 관리해보면서 운영서버는 다르게 해도 되지 않을까?

`docker-compose-dev.yml`을 EC2 인스턴스 내에서 관리한다.
- 수정하기 쉬움 (깃허브에서 변경하지 않아도 됨)
- 해당 파일 변경 내역이 남지 않아서 디버깅이 어려울 수 있음

`docker-compose-dev.yml`을 깃허브로 버전 관리한다. 환경변수는 어디서든 주입하면 되니
- 수정하기 어려움 (docker compose 파일만 수정해야 하는 경우에도 코드 병합 과정 수행해야 함)
- docker-compose 파일이 관리되며, 코드 리뷰 사이클에 포함시킬 수 있음

YehyeokBang reviewed (commented):
# 컨테이너 환경으로 변경한다면...

기존에는 각 서버에 Java, OpenTelemetry Agent 등 실행 환경을 수동으로 설치해야 했음
그래서 서버 교체나 확장 시 동일한 환경을 맞추기가 번거로웠고, 무중단 배포(Blue-Green, Rolling 등) 같은 전략을 적용하기에도 어려웠을 것임 (매번 환경을 구축하거나 그 차이를 인지하고 있어야 하기 때문)

애플리케이션을 Docker 이미지 기반으로 패키징하여 서버에 별도의 의존성 설치 없이 컨테이너 단위로 실행할 수 있도록 함

## 이미지 태그

현재는 커밋 해시(github.sha)를 이미지 태그로 사용하여 각 배포 버전을 쉽게 추적하고, 특정 커밋으로의 롤백을 할 수 있는 토대를 마련했음

다만, `태그만으로 빌드 시점을 직관적으로 파악하기 어렵`기 때문에
향후에는 날짜 + 해시 조합(YYYYMMDD-hash) 형태로도 가능할 것 같음

## 참고

이미 개발 서버 내 mysql 구축할 때 docker-compose 환경을 구축해둬서 따로 설치한 건 없음

## 머지하기 전에

- 기존 직접 실행시킨 jar 종료 및 java 제거 후 머지

jaehyeon2650 commented on .github/workflows/deploy-dev-server.yml:
```diff
@@ -22,48 +31,47 @@ jobs:
-          echo "Starting server with $JAR_FILE"
-          RUNNER_TRACKING_ID=""
-
+          IMAGE_TAG=$IMAGE_TAG \
```
IMAGE_NAME도 필요!

jaehyeon2650 reviewed (commented):
구우우웃

jaehyeon2650 commented on .github/workflows/deploy-dev-server.yml:
```diff
@@ -22,48 +31,47 @@ jobs:
-          RUNNER_TRACKING_ID=""
-
+          IMAGE_TAG=$IMAGE_TAG \
           DATABASE_URL='${{ secrets.DATABASE_DEV_URL }}' \
```
<img width="1157" height="224" alt="image" src="https://github.com/user-attachments/assets/801ef08b-8726-4a80-91e5-0b3e1b422926" />
이걸로 바꿔주세요!

YehyeokBang replied:
확인

완료

YehyeokBang committed:
fix: 일반 jar 파일이 생성되지 않도록 구성

- 빌드 시 실행 가능한 jar만 생성되도록 설정합니다.

YehyeokBang committed:
refactor: OpenTelemetry agent를 이미지에 포함하도록 패키징 구조 변경

- 빌드 시 opentelemetry-javaagent.jar를 이미지에 복사하도록 변경합니다.
- 실행 시 JAVA_TOOL_OPTIONS를 통해 agent 옵션을 주입하도록 수정합니다.

YehyeokBang commented on backend/bottari/build.gradle:
```diff
@@ -13,6 +13,14 @@ java {
     }
 }
 
+bootJar {
+    enabled = true
+}
+
+jar {
+    enabled = false
+}
```
Spring Boot 프로젝트를 Gradle로 빌드할 경우, 기본적으로 두 개의 JAR 파일이 생성됨
- plain jar → 실행 불가 (라이브러리 용도)
- boot jar → 실행 가능 (Spring Boot 애플리케이션 전체 포함)

> ### 일반(plain) jar
>
> 일반 jar 파일은 Java 애플리케이션을 라이브러리 형태로 패키징한 파일입니다. 이 파일은 클래스와 리소스만 포함되어 다른 애플리케이션에서 라이브러리로 사용될 수 있습니다. 그러나 애플리케이션을 독립적으로 실행하기 위한 모든 의존성이나 실행 환경이 포함되지 않아서, 자체적으로 실행할 수 없습니다.

그러나 Dockerfile에서는 

```
COPY --from=build /app/build/libs/*.jar /app/app.jar
```
이때 Dockerfile에서 `COPY /app/build/libs/*.jar` 로 복사할 경우 두 파일이 모두 존재하면 어떤 파일을 복사해야 할지 명확하지 않아 빌드 오류가 발생할 수 있음 그래서 아예 필요한 jar만 생성되도록 구성함

### 참고

JAR 이름을 명시적으로 지정하는 방법(COPY app-0.0.1-SNAPSHOT.jar)도 있지만,
버전 변경 시 Dockerfile을 함께 수정해야 하므로 비교적 변경에 취약하지 않은 현재 방식을 선택함

YehyeokBang committed:
Merge branch 'be/develop' into refactor/#586-deploy

# Conflicts:
#	.github/workflows/deploy-dev-server.yml

YehyeokBang committed:
fix: 불필요한 파일 제거
